File: Week 2- Algorthmic warmup/3. Greatest common multiple/gcd.cpp
#include <iostream>
// #include <chrono>
// #include <cstdlib>

// int gcd_naive(int a, int b) {
//   int current_gcd = 1;
//   for (int d = 2; d <= a && d <= b; d++) {
//     if (a % d == 0 && b % d == 0) {
//       if (d > current_gcd) {
//         current_gcd = d;
//       }
//     }
//   }
//   return current_gcd;
// }

int gcd_effiecent(long long a, long long b){
  long long num,den,mod;
  num=a;den=b;mod=num%den;
  for (long long i=0;i<=a+b;++i){
    if (mod==0){
      return den;
      break;
    }
    else{
      num=den;
      den=mod;
      mod=num%den;
    }
  }
}

int main() {
  long long a, b;
   std::cin >> a >> b;

  /// Stress Test
//   int ran_num1,ran_num2;
//   while(true)
//
// { ran_num1=rand() % 1000 + 2;
//   ran_num2=rand() % 1000 + 2;
//   int res1=gcd_naive(ran_num1,ran_num2);
//   int res2=gcd_effiecent(ran_num1,ran_num2);
//   if (res1!=res2){
//     std::cout<< "Wrong Answer"<<"\n" << "Result_Naive="<<res1<<"Result_Fast="<<res2;
//     break;
//   }
//   else std::cout<<"OK";}


  // Start Time
  // auto starts = std::chrono::steady_clock::now();

  // Code
    std::cout << gcd_effiecent(a, b);

  // End Time
  // auto ends = std::chrono::steady_clock::now() ;

  //output
  // std::cout << "elapsed processor time: " << double(std::chrono::duration_cast <std::chrono::nanoseconds> (ends-starts).count())<< " nsecs\n";
  return 0;
}

File: Week 2- Algorthmic warmup/3. Greatest common multiple/gcd.py
# Uses python3
import sys
# import time
# import numpy as np

def gcd_naive(a, b):
    current_gcd = 1
    for d in range(2, min(a, b) + 1):
        if a % d == 0 and b % d == 0:
            if d > current_gcd:
                current_gcd = d
    return current_gcd

def gcd_effiecent(a, b):
    num=a
    den=b
    mod=num%den
    for x in range(a+b):
        if (mod==0):
            return den
            break
        else:
            num=den
            den=mod
            mod=num%den


if __name__ == "__main__":
    # ## Stress Test
    # while (True):
    #     ran_num1=np.random.randint(1,2*10E5, dtype='int64')
    #     ran_num2=np.random.randint(1,2*10E5, dtype='int64')
    #     print(ran_num1,ran_num2)
    #     res1=gcd_effiecent(ran_num1,ran_num2)
    #     res2=gcd_naive(ran_num1,ran_num2)
    #     if (res1!=res2):
    #         print("Wrong Answer \n fast=",res1,"and naive=",res2)
    #         break
    #     else: print("OK")
    ## Code

    input = sys.stdin.read()
    # start_time = time.time()
    a, b = map(int, input.split())
    # print(gcd_naive(a, b))
    # end_time=time.time()
    # print("--- %s seconds ---" % (end_time - start_time))
    print(gcd_effiecent(a, b))
    # print("--- %s seconds ---" % (time.time() - end_time))
